### PR TITLE
inductor dtype propagation fixes

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141495

- Add in upcast_compute_type on creation of new tensors (loads, constants)
- Fixes index_expr - right now we are sort of inconsistent in dtype and dont always respect the dtype specified. would be nice to fix but not doing in this pr.
- bug fix in view dtype where we were always upcasting back to fp32 when input was in bf16/fp16. we should only be doing that if the output is also in bf16/fp16.
- for masked, avoid calling dtype propagation and just use output dtype.

Turns on the runtime dtype verification for opinfo tests. The separate test file is still useful because we can use it for testing turning off codegen_upcast_to_fp32.

Follow ups:

- We could consider requiring less explicit upcast_compute_types calls and do it automatically. That would potentially make things easier but be less flexible in the future. Maybe I should have done it this pr.
- Be more consistent on our index expr dtype printing.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -1972,6 +1972,7 @@ def _compile_to_module(self) -> ModuleType:
         try:
             linemap = [(line_no, node.stack_trace) for line_no, node in linemap]  # type: ignore[misc]
             key, path = PyCodeCache.write(code)
+            output_code_log.debug("Output code written to: \n%s", path)
```
Comment (Skylion007) [inline torch/_inductor/graph.py:1975]: You want a new line before the path?

File: torch/_inductor/dtype_propagation.py [hunk 2]
```diff
@@ -138,28 +153,42 @@ def return_dtype(*args: DTypeArg, dtype: torch.dtype) -> torch.dtype:
...
+        return upcast_compute_type(V.graph.get_dtype(loads[-1].args[1]))
 
     @staticmethod
     def where(a: DTypeArg, b: DTypeArg, c: DTypeArg) -> torch.dtype:
         return promote_types([b, c])
 
     @staticmethod
     def index_expr(expr: sympy.Expr, dtype: torch.dtype) -> torch.dtype:
-        return dtype
+        # TODO - rationalize index_expr. the dtype is not always used and we inconsistent about int32 or int64
```
Comment (blaine-rister) [inline torch/_inductor/dtype_propagation.py:181]:
```suggestion
        # TODO - rationalize index_expr. The dtype is not always used and we are inconsistent about int32 or int64
```

File: torch/_inductor/utils.py [hunk 3]
```diff
@@ -2247,6 +2247,15 @@ def register_op_dtype_propagation_rules(
     )
 
 
+def upcast_compute_type(dtype: torch.dtype) -> torch.dtype:
+    """Maybe upcast [b]float16 to float32"""
+    if config.triton.codegen_upcast_to_fp32 and (
+        dtype == torch.float16 or dtype == torch.bfloat16
```
Comment (blaine-rister) [inline torch/_inductor/utils.py:2253]: very nit. Just my personal preference. Up to you.
```suggestion
        dtype in {torch.float16, torch.bfloat16}
```

**Review by blaine-rister — APPROVED** [review]
LGTM! Left a few tiny style nits.

**Comment by blaine-rister** [comment]
nit: Would it make sense to update the title to mention dtype propagation? I'm not familiar with ghstack and whether each stacked PR will show up in the commit history separately.

**Review by arui-meta — APPROVED** [review]
🔥 LGTM!

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**